docs/en/observability/monitor-k8s/monitor-k8s-application-performance.asciidoc

```diff
@@ -0,0 +1,304 @@
+Useful when all pods in a node should share a single APM Server instance.
+
+* Deploy APM Server as a sidecar -- For environments that should not share an APM Server,
+like when directing traces from multiple applications to separate {es} clusters.
+
+* {apm-server-ref-v}/installing.html[Download and install APM Server] -- The classic, non-Kubernetes option.
+
+[discrete]
+=== Step 2: Set up APM Agents
+
+In most cases, setting up APM agents and thereby instrumenting your applications
+is as easy as installing a library and adding a few lines of code.
+See the relevant agent setup instructions for more information:
+
+* {apm-go-ref-v}/getting-started.html[Go agent]
+* {apm-java-ref-v}/setup.html[Java agent]
+* {apm-dotnet-ref-v}/setup.html[.NET agent]
+* {apm-node-ref-v}/set-up.html[Node.js agent]
+* {apm-py-ref-v}/set-up.html[Python agent]
+* {apm-ruby-ref-v}/set-up.html[Ruby agent]
+
+[discrete]
+=== Step 3: (Optional) Attach the agent
+
+Some agents offer the ability to instrument supported technologies without
+any changes to an application image or code.
+To do this, you'll need an https://kubernetes.io/docs/concepts/workloads/pods/init-containers/[init container]
+based on an official Elastic APM docker image.
+
+Before your application starts, copy the agent from the init container into a shared volume.
+For example, with the Java agent:
+
+[source,yml]
+----
+    # ...
+    spec:
+      volumes:
+      - name: elastic-apm-agent <1>
+        emptyDir: {}
+      initContainers:
+      - name: elastic-java-agent
+        image: docker.elastic.co/observability/apm-agent-java:1.12.0 <2>
+        volumeMounts:
+        - mountPath: /elastic/apm/agent
+          name: elastic-apm-agent
+        command: ['cp', '-v', '/usr/agent/elastic-apm-agent.jar', '/elastic/apm/agent'] <3>
+----
+<1> The shared volume
+<2> Build the initContainer from the official Elastic Java agent image
+<3> Copy the agent to the shared volume
+
+The Java command line needs a way to pick up this `javaagent` configuration.
+You can use the standard JVMTI https://docs.oracle.com/javase/8/docs/platform/jvmti/jvmti.html#tooloptions[JAVA_TOOL_OPTIONS] environment variable to do this.
+It doesn't have to be explicitly specified and is picked up automatically by the JVM when it starts.
+
+TIP: For JVMs that don't support this option, you can use any other environment variable --
+either one already defined in your startup script, like `JAVA_OPTS` in some servlet container scripts,
+or add a dedicated empty one that will have no effect if it's not set.
+
+[source,yml]
+----
+      # ...
+      containers:
+      - name: your-app-container
+        env:
+        # ...
+        - name: JAVA_TOOL_OPTIONS <1>
+          value: -javaagent:/elastic/apm/agent/elastic-apm-agent.jar
+----
+<1> Used for the command line to pick up the `javaagent` configuration
+
+[discrete]
+=== Step 4: Save the secret token
```

Comment:
None of the configuration done in the previous steps can be done without having the k8s secrets created, so "Save the secret token" and "configure the agent" should move up to steps 2/3.  Here is a fully worked out example: https://github.com/elastic/examples/tree/master/APM/Node.js/Kubernetes

Take a look at how simple it is to use APM in k8s (based on above example), and then take a look at the docs the way they are written.  I think you will see that pointing the reader to the agent docs is a problem.  If, for example, they look at the node agent docs and add the block to their code they will have no idea what to put in there, right?  What they need to do really is setup the environment vars via k8s secrets, and then follow the advanced instructions in the node agent docs instead of the basic instructions (and I think the advanced instructions are actually easier than the basic ones).

I think I would approach this like so:

Configuring your application for Elastic APM is done by:
 - Gathering configuration parameters from your APM server
 - Setting environment variables in your pods via k8s secrets
 - Adding the Elastic APM library (jar file, Node.js package, etc.) to your pods
 - Configuring your application to include the APM library

The first two steps above could be in this observability doc, and the last two could be in the agent docs.  Using the Node agent doc as an example, the Advanced config ( https://www.elastic.co/guide/en/apm/agent/nodejs/current/express.html#express-advanced-configuration) could be renamed to "Configuration via environment variables or Kubernetes secrets", and in the body it could say "If you set these environment vars, simply add this one line to the top of your application’s main file (usually index.js, server.js or app.js)."

Reply:
Really great feedback here. I've addressed it in #316 and 10a4ea8.

> I think you will see that pointing the reader to the agent docs is a problem. If, for example, they look at the node agent docs and add the block to their code they will have no idea what to put in there, right?

> The first two steps above could be in this observability doc, and the last two could be in the agent docs.

Instead of linking to the agent docs as you suggested, I ended up pulling the relevant agent content into this guide. I'm curious to hear your feedback on this approach.